Backport: Removing hidden/reserved roles added via rolesmapping

debjanibnrj opened the pull request:
Earlier the InternalUsers API was filtering out all hidden/reserved roles. We will just restrict it to hidden/reserved roles associated via rolesmapping. This will ensure behaviour of rolesmapping and internalusers API is the same.

(cherry picked from commit 6c1f002b25d77f2636c86b419dc4be7a00f889fe)

*Issue #, if available:*

*Description of changes:*


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

hardik-k-shah commented on src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/InternalUsersApiAction.java:
```diff
@@ -115,18 +115,23 @@ protected void handlePut(RestChannel channel, final RestRequest request, final C
         final SecurityJsonNode securityJsonNode = new SecurityJsonNode(contentAsNode);
 
-        // Don't allow user to add hidden, reserved or non-existent role
+        // Don't allow user to add hidden, reserved or non-existent rolesmapping
```
more explicit comment can be  " Don't allow user to add non-existent role or a role for which ole-mapping is hidden or reserved "

hardik-k-shah commented on src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/InternalUsersApiAction.java:
```diff
@@ -115,18 +115,23 @@ protected void handlePut(RestChannel channel, final RestRequest request, final C
 
-                if (isReadOnly(rolesConfiguration, role)) {
+                if (isReadOnly(rolesmappingConfiguration, role)) {
                     forbidden(channel, "Role '" + role + "' is read-only.");
```
role + "' has read-only role-mapping");

hardik-k-shah commented on src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/InternalUsersApiAction.java:
```diff
@@ -115,18 +115,23 @@ protected void handlePut(RestChannel channel, final RestRequest request, final C
+                    return;
+                }
+                if (isHidden(rolesmappingConfiguration, role)) {
+                    notFound(channel, "Role '"+role+"' is not available.");
```
+role+"' is not available for role-mapping.")

hardik-k-shah commented on src/test/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/UserApiTest.java:
```diff
@@ -123,7 +123,7 @@ public void testUserApi() throws Exception {
-        Assert.assertEquals(settings.get("message"), "Role 'opendistro_security_hidden' is not found.");
+        Assert.assertEquals(settings.get("message"), "Role 'opendistro_security_hidden' is not available.");
 
         // Associating with reserved role is allowed (for superamdin)
```
should we add similar test to verify that non-superadmin cannot add roles if roles-mapping is reserved

debjanibnrj replied:
There is an existing test to check if non-superadmin cannot add reserved rolesmapping -  https://github.com/opendistro-for-elasticsearch/security/blob/master/src/test/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/UserApiTest.java#L561-L566

hardik-k-shah commented on src/test/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/UserApiTest.java:
```diff
@@ -123,7 +123,7 @@ public void testUserApi() throws Exception {
             new Header[0]);
```
on line 121 comment 
Associating with  role is not allowed if roles-mapping is hidden

debjanibnrj commented:
> We should not allow non-superadmin to update user or rolemapping if
> (1) role is hidden
> (2) if role mapping is hidden
> (3) if role mapping is reserved
> 
> #2 and #3 are covered.
> 
> Any thoughts on #1 --> If role is hidden, that means superadmin doesn't want anyone to see that role and hence we should not allow anyone to update role-mapping or role-mapping field (for a hidden role)

Added #1 as part of the following issue https://github.com/opendistro-for-elasticsearch/security/issues/590. I will need to update both the rolesmapping and internalusers API to accomodate this use case.

debjanibnrj commented:
I have addressed comments for this review as part of https://github.com/opendistro-for-elasticsearch/security/pull/614 (@hardik-k-shah). Will close this PR and address comments in #614 before backporting.
